compiler/noirc_frontend/src/ownership/tests.rs

```diff
@@ -0,0 +1,118 @@
+#![cfg(test)]
+//! The easiest way to test this pass is a bit indirect. We have to run
+//! ownership in its entirety and look at where the clones are inserted.
+//! Testing e.g. the last_use pass directly is difficult since it returns
+//! sets of IdentIds which can't be matched to the source code easily.
+
+use crate::test_utils::get_monomorphized_no_emit_test;
+
+#[test]
+fn last_use_in_if_branches() {
+    let src = "
+    unconstrained fn main(d: [Field; 2]) {
+        if len(d) == 2 {              // use 1 of d
+            if len(d) == 2 {          // use 2 of d
+                assert(eq(d, [5, 6]));  // use 3 of d
+            }
+        } else {
+            assert(eq(d, [5, 6]));      // use 4 of d
+        }
+    }
+
+    fn eq(lhs: [Field; 2], rhs: [Field; 2]) -> bool {
+        (lhs[0] == rhs[0]) & (lhs[1] == rhs[1])
+    }
+    
+    fn len(arr: [Field; 2]) -> u32 {
+        2
+    }
+    ";
+
+    let expected = "unconstrained fn main$f0(d$l0: [Field; 2]) -> () {
+    if (len$f1(d$l0.clone()) == 2) {
+        if (len$f1(d$l0.clone()) == 2) {
+            constrain eq$f2(d$l0, [5, 6]);
+        }
+    } else {
+        constrain eq$f2(d$l0, [5, 6]);
+    }
+}
+unconstrained fn len$f1(arr$l1: [Field; 2]) -> u32 {
+    2
+}
+unconstrained fn eq$f2(lhs$l2: [Field; 2], rhs$l3: [Field; 2]) -> bool {
+    ((lhs$l2[0] == rhs$l3[0]) & (lhs$l2[1] == rhs$l3[1]))
+}
+";
+
+    let program = get_monomorphized_no_emit_test(src).unwrap();
+    similar_asserts::assert_eq!(expected, program.to_string());
```

Comment:
At the risk of being a broken record... `insta::assert_snapshot!` can be used in much the same way but makes it easier to update these in future if necessary.

Reply:
I don't really like how it makes it so that you can't see what `expected` code you're testing for. It makes it unclear from the source code what cases are actually covered. If you'd still prefer it I'll add it though.

Reply:
Ok, thank you. I was getting odd errors with `insta::assert_snapshot!(expected, program.to_string());` expecting a file name and must have gotten confused.